### ISSUE TITLE
Markdown to Pdf (via latex). Minimal custom template breaks if markdown file contains a header.

### ISSUE THREAD

**Issue by JohnLukeBentley** [jgm/pandoc#7264]
When using pandoc to transform from markdown to pdf (via latex), and using a minimal custom template, it breaks if there's at least one heading in the body.

To reproduce:
TestMinimal.md
```markdown
---
title: Test Minimal
...

## Volutpat

Lorem ipsum dolor sit amet consectetuer volutpat facilisi malesuada vitae et. Pulvinar volutpat ipsum.
```
minimal.latex (a custom pandoc template):
```latext
\documentclass{article}
$if(title)$
\title{$title$}
$endif$
\begin{document}
\maketitle
$body$
\end{document}
```

The command:
> pandoc TestMinimal.md --from markdown --to pdf --output .\output\TestMinimal.pdf --template .\minimal.latex --pdf-engine=pdflatex --standalone

The result:
> Error producing PDF.
! Undefined control sequence.
l.6 \hypertarget

I debug this by outputting to an intermediate tex (latex) file with: 
> pandoc TestMinimal.md --from markdown --to latex --output .\output\TestMinimal.tex --template .\minimal.latex --pdf-engine=pdflatex --standalone

The resulting TestMinimal.tex:
```latex
\documentclass{article}
\title{Test Minimal}
\begin{document}
\maketitle
\hypertarget{volutpat}{%
\subsection{Volutpat}\label{volutpat}}

Lorem ipsum dolor sit amet consectetuer volutpat facilisi malesuada
vitae et. Pulvinar volutpat ipsum.
\end{document}
```

If, in my preferred latex UI (TexStudio), I try to compile this I get more or less the same error
> ! Undefined control sequence.
l.5 \hypertarget
                {volutpat}{%
The control sequence at the end of the top line of your error message was never \def'ed. If you have misspelled it (e.g., `\hobx'), type `I' and the correct spelling (e.g., `I\hbox'). Otherwise just continue, and I'll forget about whatever was undefined.

If I edit TestMinimal.tex to include the hyperref package ...
```latex
\documentclass{article}
\title{Test Minimal}
\usepackage{hyperref}
\begin{document}
\maketitle
\hypertarget{volutpat}{%
\subsection{Volutpat}\label{volutpat}}

Lorem ipsum dolor sit amet consectetuer volutpat facilisi malesuada
vitae et. Pulvinar volutpat ipsum.
\end{document}
```
... I can now compile TestMinimal.tex  successfully into a pdf file.

It appears, therefore, a `\hypertarget` command requires the `hyperref` package. On the assumption that is true then ...

The problem appears to be: pandoc wrongly inserts, into the .tex intermediate file,  a `\hypertarget` command even when no `hyperref` package has been defined in the pandoc .latex template.

Presumably the solution is (wherever the relevant pandoc code lives) a simple `if ... then` check to only insert a `\hypertarget` command if the .latex template has the `hyperref` package defined. Otherwise insert a plain `\subsection{Volutpat}`. 

pandoc version 2.13

**Comment by tarleb** [comment]
The high level view of what pandoc does when outputting LaTeX looks like this:

1. Convert internal document representation to LaTeX;
2. Set variables noting which document features are required for rendering (e.g.: table or no table);
3. Process template using the given variables; insert the produced LaTeX snippets.

So the processing of the template happens only after the LaTeX code has been generated. Furthermore, pandoc does not parse the template and check for available features, but just replaces variables.

In other words: the expectation is a bit backwards here. Templates must provide all features to ensure that the generated code can be compiled correctly, not the other way around.

The hyperref commands are created if the heading as an identifier. The id is automatically assigned while reading Markdown, but this feature can be turned off by using `--from=markdown-auto_identifiers`. Using that should solve your problem.

**Comment by tarleb** [comment]
[This tex.SE answer](https://tex.stackexchange.com/q/480427/148113) may be helpful, too.

**Comment by jgm** [comment]
@tarleb is right.  You are responsible for including in the template everything that pandoc might require.  Starting from pandoc's own default template as a base is a good idea.

**Comment by JohnLukeBentley** [comment]
Thanks very much for the replies @tarleb and @jgm.

Ill take the liberty of trying to bed down an understanding of why it's warranted to close down these sort of issues (together with #7265). The length of this post might be endured as something which, I hope, enables any response to be short. But, in any case, neither of you should feel obliged to read on. I suspect I'm possessed of the sufficient reasons for wanting to close the issue.

My understanding of the possible strategies when using pandoc for pdf output (perhaps especially given a markdown input document) is, more or less as [@tarleb well and helpfully describes](https://tex.stackexchange.com/a/480504/105123) in the linked *SE.Tex >What is the minimal “API” a pandoc template must provide*. In my own framing ...

As a first approach attempt to rely on pandoc's default template (default.latex) with a desired combination of the many available pandoc variables/options. Pandoc variables/options sent to pandoc via some combination of: command line switches; input document yaml header; a separate file inserted with `--include-in-header`; a separate file that is processed via `--metadata-file`; a separate file that is processed via `--defaults`. Having a separate file especially allows for separating presentation from content (a hard won lesson from the web world). And the Pandoc variables/options, with the default template, are mature enough to handle a great range of typical cases.

However, if this first approach can't give you what you want then pandoc provides a means to use a custom template (e.g. my.latex), with the `--template` switch. On the face of it there's at least the following possible strategies:

1. (As you both promote) Copy the default.latex template and modify it, taking one of the following approaches:

   1.  By aggressively pairing back (latex) packages and commands that are irrelevant to one's situation. As @tarleb writes by way of example "One would likely remove everything that deals beamer presentations ....  commit to a specific LaTeX engine by dropping all xetex/luatex checks, and hard-code all packages that one intents to use"; or 
   2. More modestly add one or two (latex) commands/packages, perhaps overriding things in earlier parts of the template; or
   3. As in step two but attempting to keep separate those "modest" additions by holding them in a separate file and inserting them late into the latex template. E.g through creating a custom package of one's own (as a *.sty file) and referencing the package in the latex template with `\usepackage[]{mypackage}`. Given latex can be finicky about the order of packages this strategy might turn out to be impossible in some cases, at least in a pure form.

2. Start fresh with a minimal custom template; and, as @tarleb describes, "Extend and modify that template as you hit an issue".

I note that any of the three alternate options under 1) "Copy the default.latex template and modify it" might benefit from the [documented recommendation](https://pandoc.org/MANUAL.html#templates) to:
> [track] the changes in the default templates, and [modify] your custom templates accordingly. An easy way to do this is to fork the pandoc-templates repository and merge in changes after each pandoc release.

You'll both understand that my original post, and  #7265, was motivated by an attempt to see if pandoc might be modified to better support 2) "Start fresh with a minimal custom template". 

@tarleb seems to suggest that modifying pandoc to that end is not possible (given the internal pipeline of pandoc processing): "pandoc does not parse the template and check for available features, but just replaces variables". Without myself being familiar the relevant pandoc internals I don't know enough to contradict that. However, on an intuitive level "3. Process template using the given variables; insert the produced LaTeX snippets into the template." seems to describe a step where pandoc *is* parsing the template (which it must do to insert variables and Latex snippets) and could therefore, in principle, conditionally include or exclude output in the latex body given the presence or absence of packages in the body.

However, even on the assumption I'm right - that pandoc could in principle be modified to produce cleaner latex in the body given the state of what it finds the template header - both of you may well have in mind other reasons for rejecting my proposal. 

Perhaps that pandoc is carefully developed to work with the default templates in mind. And trying to modify pandoc to work with both the default templates and minimal custom templates could well create a development scenario as if trying to make a carpet too large for a room fit in all corners. You push pandoc down the corner of accommodating minimal custom templates and that raises the default template corner.

Perhaps more importantly, if pandoc was modified to accommodate 2) "Start fresh with a minimal custom template", e.g. by not outputting, from the start, impure commands in the body like `\hypertarget`, it wouldn't be long before the user of a minimal custom template discovers, as they turn their  minimal custom template into something more production worthy, will want the more complex commands, like `\hypertarget`, right back in the body. E.g. When they want a table of contents with clickable links to headers. In other words it's not going to be long before such a user is after all the "impure" commands in the body that pandoc currently provides and carefully handles with the default.latex template. And for this reason any attempt to alter pandoc to produce a "pure" latex body quickly works at cross purposes.

Does that, more or less, capture the reasons/concerns either of you have in mind?

(@tarleb I appreciate the `--from=markdown-auto_identifiers` tip. That would be useful in this specific case. But, as you likely subsequently appreciated (given your link to SE.tex), I was motivated to address the issue of having pandoc accommodate minimal custom templates generally).
